### PR TITLE
Fix #5888: Disallow SAM type creation of final and sealed classes

### LINKED ISSUE #5888: `sealed` and `final` classes can be extended using SAM
This is pretty straight forward.

For example:
```scala
  val x: String <:< Int = _.toInt
  println(x("3") - 2)
```
Compiles and prints 1.

### PR DESCRIPTION
**Pull request by odersky**

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 1]
```diff
@@ -985,7 +985,11 @@ class Typer extends Namer
               case SAMType(sam)
               if !defn.isFunctionType(pt) && mt <:< sam =>
                 if (!isFullyDefined(pt, ForceDegree.all))
-                  ctx.error(ex"result type of closure is an underspecified SAM type $pt", tree.sourcePos)
+                  ctx.error(ex"result type of lambda is an underspecified SAM type $pt", tree.sourcePos)
+                else if (pt.classSymbol.is(FinalOrSealed)) {
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala]: Should we move this check into the `SAMType` extractor?

**Comment by odersky** [comment]
Even anonymous subclasses of sealed types are problematic. So I think it's OK to not allow SAM type creation.

File: compiler/src/dotty/tools/dotc/typer/Typer.scala (hunk 1 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/Typer.scala]: No I think it's better here. An extractor is for extracting, not reporting errors.

Comment (allanrenucci) [reply compiler/src/dotty/tools/dotc/typer/Typer.scala]: I didn't mean to report an error in the extractor, just have the extractor return `None` on `sealed` and `final` classes. It seemed arbitrary to have a special error message for this. In `scalac` I get a type mismatch instead.:
```scala
scala> sealed trait Foo { def f(x: Int): Int }; val foo: Foo = (x: Int) => x
<console>:11: error: type mismatch;
 found   : Int => Int
 required: Foo
       sealed trait Foo { def f(x: Int): Int }; val foo: Foo = (x: Int) => x
```

**Pull request merged**